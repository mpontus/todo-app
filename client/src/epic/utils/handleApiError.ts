import { OperatorFunction, of } from "rxjs";
import { Action } from "redux";
import { catchError } from "rxjs/operators";
import { RequestError } from "../../model/RequestError";

/**
 * Helper function to handle API error
 *
 * Wraps error caught from source observable into RequestError and
 * passes it onto specified action creator before merging it into the
 * result stream.
 */
export const handleApiError = <T, R>(
  failureActionCreator: (e: any) => R
): OperatorFunction<T, T | R> =>
  catchError(error =>
    of(failureActionCreator(RequestError.fromApiError(error)))
  );
