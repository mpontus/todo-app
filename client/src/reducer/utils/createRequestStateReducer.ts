import { Reducer } from "redux";
import { Action } from "../../action";

interface RequestState<ErrorType> {
  loading: boolean;
  success: boolean;
  error?: ErrorType;
}

const initialState = {
  loading: false,
  success: false
};

export const createRequestStateReducer = <
  BaseAction extends Action,
  RequestAction extends BaseAction,
  SuccessAction extends BaseAction,
  FailureAction extends BaseAction,
  ErrorType extends Error
>(
  requestActionPredicate: (value: BaseAction) => value is RequestAction,
  successActionPredicate: (value: BaseAction) => value is SuccessAction,
  failureActionPredicate: (value: BaseAction) => value is FailureAction,
  errorSelector: (value: FailureAction) => ErrorType
): Reducer<RequestState<ErrorType>, BaseAction> => (
  state = initialState,
  action
) => {
  if (requestActionPredicate(action)) {
    return {
      loading: true,
      success: false
    };
  }

  if (successActionPredicate(action)) {
    return {
      loading: false,
      success: true
    };
  }

  if (failureActionPredicate(action)) {
    return {
      loading: false,
      success: false,
      error: errorSelector(action)
    };
  }

  return state;
};
