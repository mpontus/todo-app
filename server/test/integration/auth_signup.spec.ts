import supertest from 'supertest';
import { initApp } from '../utils/initApp';
import { resetDb } from '../utils/resetDb';

let expressApp: any;
let nestApp: any;

beforeAll(async () => {
  ({ expressApp, nestApp } = await initApp());
});

afterAll(() => nestApp.close());

beforeEach(resetDb);

describe('signup', () => {
  const authSeed = require('../seeds/anonymous_user');

  beforeEach(() => authSeed.run());

  describe('when the user does not exist', () => {
    it('should be successful', async () => {
      const response = await supertest(expressApp)
        .post('/auth/signup')
        .send({
          username: 'Hailee58',
          password: '9O_8ywUKpHuHjnZ',
        })
        .set('Authorization', `Bearer ${authSeed.token}`)
        .expect(201);

      expect(response.body).toMatchSnapshot({
        token: expect.any(String),
        user: expect.objectContaining({
          id: authSeed.id,
          username: 'Hailee58',
        }),
      });
    });
  });

  describe('when the username is missing', () => {
    it('should be an error', async () => {
      const response = await supertest(expressApp)
        .post('/auth/signup')
        .send({
          password: '9O_8ywUKpHuHjnZ',
        })
        .set('Authorization', `Bearer ${authSeed.token}`)
        .expect(400);

      expect(response.body).toMatchSnapshot();
    });
  });

  describe('when username is already taken', () => {
    const seed = require('../seeds/registered_user');

    beforeEach(() => seed.run());

    it('should be an error', async () => {
      const response = await supertest(expressApp)
        .post('/auth/signup')
        .send({
          name: seed.username,
          password: '9O_8ywUKpHuHjnZ',
        })
        .set('Authorization', `Bearer ${authSeed.token}`)
        .expect(400);

      expect(response.body).toMatchSnapshot();
    });
  });

  describe('when the username is too long', () => {
    it('should be an error', async () => {
      const response = await supertest(expressApp)
        .post('/auth/signup')
        .send({
          username: 'x'.repeat(1000),
          password: '9O_8ywUKpHuHjnZ',
        })
        .set('Authorization', `Bearer ${authSeed.token}`)
        .expect(400);

      expect(response.body).toMatchSnapshot();
    });
  });

  describe('when the password is missing', () => {
    it('should be an error', async () => {
      const response = await supertest(expressApp)
        .post('/auth/signup')
        .send({
          username: 'Hailee58',
        })
        .set('Authorization', `Bearer ${authSeed.token}`)
        .expect(400);

      expect(response.body).toMatchSnapshot();
    });
  });

  describe('when the password is too short', () => {
    it('should be an error', async () => {
      const response = await supertest(expressApp)
        .post('/auth/signup')
        .send({
          username: 'Hailee58',
          password: 'foo',
        })
        .set('Authorization', `Bearer ${authSeed.token}`)
        .expect(400);

      expect(response.body).toMatchSnapshot();
    });
  });
});
