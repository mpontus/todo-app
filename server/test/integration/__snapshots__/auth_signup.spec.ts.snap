// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`signup when the password is missing should be an error 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    Object {
      "children": Array [],
      "constraints": Object {
        "minLength": "password must be longer than or equal to 6 characters",
      },
      "property": "password",
      "target": Object {
        "username": "Hailee58",
      },
    },
  ],
  "statusCode": 400,
}
`;

exports[`signup when the password is too short should be an error 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    Object {
      "children": Array [],
      "constraints": Object {
        "minLength": "password must be longer than or equal to 6 characters",
      },
      "property": "password",
      "target": Object {
        "password": "foo",
        "username": "Hailee58",
      },
      "value": "foo",
    },
  ],
  "statusCode": 400,
}
`;

exports[`signup when the user does not exist should be successful 1`] = `
Object {
  "token": Any<String>,
  "user": ObjectContaining {
    "id": "9dc8c9df-74be-5ca0-93d6-b049e820f13f",
    "username": "Hailee58",
  },
}
`;

exports[`signup when the username is missing should be an error 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    Object {
      "children": Array [],
      "constraints": Object {
        "maxLength": "username must be shorter than or equal to 255 characters",
      },
      "property": "username",
      "target": Object {
        "password": "9O_8ywUKpHuHjnZ",
      },
    },
  ],
  "statusCode": 400,
}
`;

exports[`signup when the username is too long should be an error 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    Object {
      "children": Array [],
      "constraints": Object {
        "maxLength": "username must be shorter than or equal to 255 characters",
      },
      "property": "username",
      "target": Object {
        "password": "9O_8ywUKpHuHjnZ",
        "username": "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
      },
      "value": "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
    },
  ],
  "statusCode": 400,
}
`;

exports[`signup when username is already taken should be an error 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    Object {
      "children": Array [],
      "constraints": Object {
        "maxLength": "username must be shorter than or equal to 255 characters",
      },
      "property": "username",
      "target": Object {
        "name": "dodgip",
        "password": "9O_8ywUKpHuHjnZ",
      },
    },
  ],
  "statusCode": 400,
}
`;
